packages/core/test/actions.test.ts

```diff
@@ -847,6 +847,254 @@ describe('entry/exit actions', () => {
+          }
+        }
+      });
+
+      const parent = createMachine({
+        id: 'parent',
+        context: () => ({
+          child: spawn(child)
+        }),
+        on: {
+          STOP_CHILD: {
+            actions: stop((ctx: any) => ctx.child)
+          },
+          EXIT: {
+            actions: () => {
+              throw new Error('This should not be called.');
+            }
+          }
+        }
+      });
+
+      const interpreter = interpret(parent).start();
+      interpreter.send({ type: 'STOP_CHILD' });
+    });
+
+    it('actors spawned in exit handlers of a stopped child should not be started', () => {
+      const grandchild = createMachine({
+        id: 'grandchild',
+        entry: () => {
+          throw new Error('This should not be called.');
+        }
+      });
+
+      const parent = createMachine({
+        id: 'parent',
+        context: {},
+        exit: assign({
+          actorRef: () => spawn(grandchild)
+        })
+      });
+
+      const interpreter = interpret(parent).start();
+      interpreter.stop();
+    });
+
+    it('should execute referenced custom actions correctly when stopping an interpreter', () => {
+      let called = false;
+      const parent = createMachine(
+        {
+          id: 'parent',
+          context: {},
+          exit: 'referencedAction'
+        },
+        {
+          actions: {
+            referencedAction: () => {
+              called = true;
+            }
+          }
+        }
+      );
+
+      const interpreter = interpret(parent).start();
+      interpreter.stop();
+
+      expect(called).toBe(true);
+    });
+
+    it('should execute builtin actions correctly when stopping an interpreter', () => {
+      const machine = createMachine(
+        {
+          context: {
+            executedAssigns: [] as string[]
+          },
+          exit: [
+            'referencedAction',
+            assign({
+              executedAssigns: (ctx: any) => [...ctx.executedAssigns, 'inline']
+            })
+          ]
+        },
+        {
+          actions: {
+            referencedAction: assign({
+              executedAssigns: (ctx) => [...ctx.executedAssigns, 'referenced']
+            })
+          }
+        }
+      );
+
+      const interpreter = interpret(machine).start();
+      interpreter.stop();
+
+      expect(interpreter.state.context.executedAssigns).toEqual([
+        'referenced',
+        'inline'
+      ]);
+    });
+
+    it('should clear all scheduled events when the interpreter gets stopped', () => {
+      const machine = createMachine({
+        on: {
+          INITIALIZE_SYNC_SEQUENCE: {
+            actions: () => {
+              // schedule those 2 events
+              service.send({ type: 'SOME_EVENT' });
+              service.send({ type: 'SOME_EVENT' });
+              // but also immediately stop *while* the `INITIALIZE_SYNC_SEQUENCE` is still being processed
+              service.stop();
```

Comment:
So `service.stop()` takes precedence over `service.send()` events within the same "tick", if I'm understanding this right?

Is this because `service.stop()` is meant to be synchronous, and `service.send()` is (conceptually) async?

Reply:
If we just focus on the concept here then I think that my mental picture here is that `xstate.stop` is just a priority signal. And since we can have stuff in the "mailbox" while receiving this signal then we need to drop those items from the mailbox and process `xstate.stop` asap (after processing currently processed item, if applicable).

So conceptually it's more like:
- when we process an event new stuff can always arrive
- if we receive a regular event then we just append it to the mailbox
- if we receive a stop signal then we clear the mailbox and append this signal to the mailbox

So both are received and processed in an asynchronous manner. It's just that I didn't figure out a better way to push to this conceptual mailbox to represent this situation in the test code. Of course, this simplified situation isn't that realistic, but those sends and stop could be replaced with some calls that would end up calling those and that is, edge-casey, but somewhat realistic. This test case just represents a "flattened", to its bare bones, situation like this.

Reply:
That makes sense - priority signals vs. messages 👍